pkg/rpc/server/server.go

```diff
@@ -1958,24 +1978,57 @@ func (s *Server) traverseIterator(reqParams request.Params) (interface{}, *respo
 	if count > s.config.MaxIteratorResultItems {
 		return nil, response.NewInvalidParamsError(fmt.Sprintf("iterator items count is out of range (%d at max)", s.config.MaxIteratorResultItems))
 	}
-	
+
 	s.sessionsLock.Lock()
 	session, ok := s.sessions[sID.String()]
 	if !ok {
 		s.sessionsLock.Unlock()
 		return []json.RawMessage{}, nil
 	}
-	session.till = time.Now().Add(time.Second * time.Duration(s.config.SessionExpirationTime))
 	session.iteratorsLock.Lock()
+	// Perform `till` update only after session.iteratorsLock is taken in order to have more
+	// precise session lifetime.
+	session.till = time.Now().Add(time.Second * time.Duration(s.config.SessionExpirationTime))
 	s.sessionsLock.Unlock()
 
 	var (
-		iIDStr = iID.String()
-		iVals  []stackitem.Item
+		iIDStr  = iID.String()
+		iVals   []stackitem.Item
+		respErr *response.Error
 	)
-	for _, it := range session.iterators {
+	for i, it := range session.iteratorIdentifiers {
 		if iIDStr == it.ID {
-			iVals = iterator.Values(it.Item, count)
+			if len(session.iterators) == 0 {
+				// We can't easily create a DB snapshot to keep the session alive for a long time,
+				// thus, we'll use historic calls approach and MPT-based storage for iterator
```

Comment:
Which breaks for KeepOnlyLatest case.

Reply:
Yes, but we can't store original iterators due to BoltDB behaviour, it will hang until iterator is released.

Reply:
IIRC it only has write problems when it needs to remap itself, otherwise snapshots should be functional there too.

Reply:
I've search the bbolt documentation and found the following:
```
Opening a read transaction and a write transaction in the same goroutine can cause the writer to deadlock because the database periodically needs to re-mmap itself as it grows and it cannot do that while a read transaction is open.

If a long running read transaction (for example, a snapshot transaction) is needed, you might want to set DB.InitialMmapSize to a large enough value to avoid potential blocking of write transaction. 
```

So adjusting `InitialMmapSize` could help us, but we still can't be 100% sure that subsequent iterator-related request won't block the whole blockhain operation/persist. Are you sure that we can't use MPT for iterator sessions support?

If we still want to use bolt snapshots, then maybe it's better to restrict the maximum amount of opened iterator sessions.

Reply:
So do we need to add something like `MaxConcurrentSessionsCount` to the server config?

Reply:
We can add this option too, but it's not critical.

Reply:
Though maybe it'll allow for somewhat safe use of iterators for public servers.

Reply:
Oh, I've missed this comment. In progress.

Reply:
Implemented. Not sure, whether we need a special value for `SessionPoolSize` that means `no constraint is set`.